### PR TITLE
Synchronize document text updates

### LINKED ISSUE #793: Application of incremental text changes is not atomic?
In SublimeText we have this "format on save" feature. When it is enabled, and when the document is saved by the user, a textDocument/didChange is notified, and immediately after a textDocument/formatting is requested.

It can sometimes happen that pyls returns text edits that destroy / chop up my python file.

I'm looking at the code in pyls and it looks like a thread pool of some sort is used to handle requests and notifications. The textDocument/didChange must be atomic in this case, i.e. other requests/notifications must wait for the handler for textDocument/didChange to finish. I'm unsure whether I'm looking in the right corner of the codebase.

### PR DESCRIPTION
**Pull request by andfoy**
Fixes #793
Closes #798

### CODE REVIEW

File: pyls/workspace.py [hunk 1]
```diff
@@ -15,6 +17,14 @@
 RE_END_WORD = re.compile('^[A-Za-z_0-9]*')
 
 
+def lock(method):
```
Comment (ccordoba12) [inline pyls/workspace.py:20]: Please add a docstring to this function.

File: pyls/workspace.py [hunk 2]
```diff
@@ -218,6 +232,7 @@ def word_at_position(self, position):
 
         return m_start[0] + m_end[-1]
 
+    @lock
     def jedi_names(self, all_scopes=False, definitions=True, references=False):
```
Comment (ccordoba12) [inline pyls/workspace.py:237]: Why does `jedi_names` need to be locked, but not the other Jedi operations?

**Review by ccordoba12 — COMMENTED** [review]
Thanks @andfoy for this! Could you mention the motivation for these changes in the PR description?

Also, is it possible to add a test for this?

File: pyls/workspace.py (hunk 2 above)
Comment (ccordoba12) [reply pyls/workspace.py:237]: Since you added the `lock` decorator to `jedi_script`, does the one above this line is still necessary?

**Review by ccordoba12 — COMMENTED** [review]
@andfoy, is it possible to add a test for this?

**Comment by andfoy** [comment]
It is not possible to guarantee a certain scenario for concurrent accesses, since it is not possible to ensure that a specific thread arrives before other

File: pyls/workspace.py (hunk 2 above)
Comment (ccordoba12) [reply pyls/workspace.py:237]: What do you think about what I said in my comment above? Do you really need to add the `lock` decorator to `jedi_names`?

**Review by ccordoba12 — APPROVED** [review]
Thanks @andfoy, looks good to me.

**Pull request merged**